File: Kotlin/src/se/devies/tollcollector/TollFreeVehiclesTest.kt
package se.devies.tollcollector

import se.devies.tollcollector.holiday.MockHolidayChecker
import org.junit.Test

import org.junit.Assert.*
import se.devies.tollcollector.vehicle.*
import java.util.*

// I have made the assumption that a se.devies.tollcollector.vehicle can only be of a single type,
// i.e., no "emergency tractors" etc.
class TollFreeVehiclesTest {
    private val tollCalculator = TollCalculator(MockHolidayChecker())

    @Test
    fun carsShouldPayToll() {
        val car = Car()

        val testCases: ArrayList<GregorianCalendar> = arrayListOf(
            GregorianCalendar(2019, 3, 1, 7, 0),
            GregorianCalendar(2019, 3, 1, 8, 10),
            GregorianCalendar(2019, 3, 1, 9, 59),
            GregorianCalendar(2019, 3, 1, 12, 30),
            GregorianCalendar(2019, 3, 1, 15, 30),
            GregorianCalendar(2019, 3, 1, 16, 59)
        )

        testCases.forEach{ date -> assertNotEquals(0, tollCalculator.getTotalTollFeeForOneDay(car, date)) }
    }

    @Test
    fun diplomatVehiclesShouldNotPayToll() {
        val vehicle = Diplomat()

        val testCases: ArrayList<GregorianCalendar> = arrayListOf(
            GregorianCalendar(2019, 3, 1, 7, 0),
            GregorianCalendar(2019, 3, 1, 8, 10),
            GregorianCalendar(2019, 3, 1, 9, 59),
            GregorianCalendar(2019, 3, 1, 12, 30),
            GregorianCalendar(2019, 3, 1, 15, 30),
            GregorianCalendar(2019, 3, 1, 16, 59)
        )

        testCases.forEach{ date -> assertEquals(0, tollCalculator.getTotalTollFeeForOneDay(vehicle, date)) }
    }

    @Test
    fun emergencyVehiclesShouldNotPayToll() {
        val vehicle = Emergency()

        val testCases: ArrayList<GregorianCalendar> = arrayListOf(
            GregorianCalendar(2019, 3, 1, 7, 0),
            GregorianCalendar(2019, 3, 1, 8, 10),
            GregorianCalendar(2019, 3, 1, 9, 59),
            GregorianCalendar(2019, 3, 1, 12, 30),
            GregorianCalendar(2019, 3, 1, 15, 30),
            GregorianCalendar(2019, 3, 1, 16, 59)
        )

        testCases.forEach{ date -> assertEquals(0, tollCalculator.getTotalTollFeeForOneDay(vehicle, date)) }
    }

    @Test
    fun foreignVehiclesShouldNotPayToll() {
        val vehicle = Foreign()

        val testCases: ArrayList<GregorianCalendar> = arrayListOf(
            GregorianCalendar(2019, 3, 1, 7, 0),
            GregorianCalendar(2019, 3, 1, 8, 10),
            GregorianCalendar(2019, 3, 1, 9, 59),
            GregorianCalendar(2019, 3, 1, 12, 30),
            GregorianCalendar(2019, 3, 1, 15, 30),
            GregorianCalendar(2019, 3, 1, 16, 59)
        )

        testCases.forEach{ date -> assertEquals(0, tollCalculator.getTotalTollFeeForOneDay(vehicle, date)) }
    }

    @Test
    fun militaryVehiclesShouldNotPayToll() {
        val vehicle = Military()

        val testCases: ArrayList<GregorianCalendar> = arrayListOf(
            GregorianCalendar(2019, 3, 1, 7, 0),
            GregorianCalendar(2019, 3, 1, 8, 10),
            GregorianCalendar(2019, 3, 1, 9, 59),
            GregorianCalendar(2019, 3, 1, 12, 30),
            GregorianCalendar(2019, 3, 1, 15, 30),
            GregorianCalendar(2019, 3, 1, 16, 59)
        )

        testCases.forEach{ date -> assertEquals(0, tollCalculator.getTotalTollFeeForOneDay(vehicle, date)) }
    }

    @Test
    fun motorbikesShouldNotPayToll() {
        val vehicle = Motorbike()

        val testCases: ArrayList<GregorianCalendar> = arrayListOf(
            GregorianCalendar(2019, 3, 1, 7, 0),
            GregorianCalendar(2019, 3, 1, 8, 10),
            GregorianCalendar(2019, 3, 1, 9, 59),
            GregorianCalendar(2019, 3, 1, 12, 30),
            GregorianCalendar(2019, 3, 1, 15, 30),
            GregorianCalendar(2019, 3, 1, 16, 59)
        )

        testCases.forEach{ date -> assertEquals(0, tollCalculator.getTotalTollFeeForOneDay(vehicle, date)) }
    }

    @Test
    fun tractorsShouldNotPayToll() {
        val vehicle = Tractor()

        val testCases: ArrayList<GregorianCalendar> = arrayListOf(
            GregorianCalendar(2019, 3, 1, 7, 0),
            GregorianCalendar(2019, 3, 1, 8, 10),
            GregorianCalendar(2019, 3, 1, 9, 59),
            GregorianCalendar(2019, 3, 1, 12, 30),
            GregorianCalendar(2019, 3, 1, 15, 30),
            GregorianCalendar(2019, 3, 1, 16, 59)
        )

        testCases.forEach{ date -> assertEquals(0, tollCalculator.getTotalTollFeeForOneDay(vehicle, date)) }
    }
}

File: test/tollFreeVehicles.test.js
const TollCalculator = require("../tollCalculator");
const Vehicle = require("../vehicle");
let tollCalculator;

describe("Toll free vehicles", () => {
  beforeAll(() => {
    tollCalculator = new TollCalculator();
  });

  it("Military vehicles are toll free", () => {
    const toll = tollCalculator.getTollFee(
      new Date("2013-01-11T15:48:00"),
      new Vehicle("Military")
    );
    expect(toll).toBe(0);
  });

  it("Tractor vehicles are toll free", () => {
    const toll = tollCalculator.getTollFee(
      new Date("2013-01-11T15:48:00"),
      new Vehicle("Tractor")
    );
    expect(toll).toBe(0);
  });

  it("Motorbike vehicles are toll free", () => {
    const toll = tollCalculator.getTollFee(
      new Date("2013-01-11T15:48:00"),
      new Vehicle("Motorbike")
    );
    expect(toll).toBe(0);
  });

  it("Emergency vehicles are toll free", () => {
    const toll = tollCalculator.getTollFee(
      new Date("2013-01-11T15:48:00"),
      new Vehicle("Emergency")
    );
    expect(toll).toBe(0);
  });

  it("Diplomat vehicles are toll free", () => {
    const toll = tollCalculator.getTollFee(
      new Date("2013-01-11T15:48:00"),
      new Vehicle("Diplomat")
    );
    expect(toll).toBe(0);
  });

  it("Foreign vehicles are toll free", () => {
    const toll = tollCalculator.getTollFee(
      new Date("2013-01-11T15:48:00"),
      new Vehicle("Foreign")
    );
    expect(toll).toBe(0);
  });

  it("Van is not toll free", () => {
    const toll = tollCalculator.getTollFee(
      new Date("2013-01-11T15:48:00"),
      new Vehicle("Van")
    );
    expect(toll).toBe(18);
  });
  it("Car is not toll free", () => {
    const toll = tollCalculator.getTollFee(
      new Date("2013-01-11T15:48:00"),
      new Vehicle("car")
    );
    expect(toll).toBe(18);
  });
});